gateway-provider-security-jwt/src/main/java/org/apache/knox/gateway/provider/federation/jwt/filter/AbstractJWTFilter.java

```diff
@@ -385,12 +387,18 @@ protected boolean validateToken(final HttpServletRequest request,
 
     if (tokenStateService != null) {
       try {
-        if (tokenIsStillValid(tokenId)) {
-          return true;
+        if (tokenId != null) {
+          if (tokenIsStillValid(tokenId)) {
+            return true;
+          } else {
+            log.tokenHasExpired(Tokens.getTokenIDDisplayText(tokenId));
+            handleValidationError(request, response, HttpServletResponse.SC_BAD_REQUEST,
+                    "Bad request: token has expired");
+          }
         } else {
-          log.tokenHasExpired(Tokens.getTokenIDDisplayText(tokenId));
+          log.missingTokenPasscode();
           handleValidationError(request, response, HttpServletResponse.SC_BAD_REQUEST,
-                                "Bad request: token has expired");
```

Comment:
It seems has though we are assuming that if token state server is enabled for a given KnoxToken service that it cannot also accept 3rd party tokens. Is this a necessary requirement? Would falling back to only using the token expiry be acceptable? It is sort of unclear how we would verify the signature given today's approach to signing keys but if we were to add support for jwks for getting the public key for verification, it seems that we could fully support 3rd party tokens as well as KnoxTokens in the same provider deployment. If we waited for the jwks support to make that work it may be harder to find and fix these.

Reply:
Again, this code is not applicable to the cases where a JWT is provided for authentication; It is only for those cases where ONLY the token ID is provided, in which case the server-managed token state is absolutely necessary since there is nothing about the ID which can be verified by itself.

Reply:
okay.